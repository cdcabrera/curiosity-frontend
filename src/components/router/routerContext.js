import { useCallback } from 'react';
import {
  useLocation,
  useNavigate as useRRDNavigate,
  useParams,
  useResolvedPath,
  useSearchParams as useRRDSearchParams
} from 'react-router-dom';
import { pathJoin, routerHelpers } from './routerHelpers';

/**
 * Return a callback for redirecting, and replacing, towards a new path, or url.
 *
 * @callback redirect
 * @param {object} options
 * @param {Function} options.useLocation
 * @param {Function} options.useResolvedPath
 * @returns {(function(*): void)|*}
 */
const useRedirect = ({
  useLocation: useAliasLocation = useLocation,
  useResolvedPath: useAliasResolvedPath = useResolvedPath
} = {}) => {
  const { pathname } = useAliasResolvedPath();
  const { hash = '', search = '' } = useAliasLocation() || {};

  /**
   * redirect
   *
   * @param {string} route
   * @returns {undefined}
   */
  return useCallback(
    route => {
      const baseName = routerHelpers.dynamicBaseName({ pathName: pathname });
      let isUrl;

      try {
        isUrl = !!new URL(route);
      } catch (e) {
        isUrl = false;
      }

      window.location.replace((isUrl && route) || `${pathJoin(baseName, route)}${search}${hash}`);
      return undefined;
    },
    [hash, pathname, search]
  );
};

/**
 * Get a route detail from router context.
 *
 * @param {object} options
 * @param {Function} options.useRedirect
 * @param {Function} options.useParams
 * @returns {{baseName: string, errorRoute: object}}
 */
const useRouteDetail = ({
  useRedirect: useAliasRedirect = useRedirect,
  useParams: useAliasParams = useParams
} = {}) => {
  const redirect = useAliasRedirect();
  const { productPath } = useAliasParams();
  const { allConfigs, configs, firstMatch } = routerHelpers.getRouteConfigByPath({ pathName: productPath });

  if (!firstMatch) {
    redirect(routerHelpers.redirectRoute.redirect);
  }

  return {
    allProductConfigs: allConfigs,
    firstMatch,
    errorRoute: routerHelpers.errorRoute,
    productGroup: firstMatch?.productGroup,
    productConfig: (configs?.length && configs) || []
  };
};

/**
 * useNavigate wrapper, apply application config context routing
 *
 * @param {object} options
 * @param {Function} options.useLocation
 * @param {Function} options.useNavigate
 * @param {Function} options.useResolvedPath
 * @returns {Function}
 */
const useNavigate = ({
  useLocation: useAliasLocation = useLocation,
  useNavigate: useAliasNavigate = useRRDNavigate,
  useResolvedPath: useAliasResolvedPath = useResolvedPath
} = {}) => {
  const { search, hash } = useAliasLocation();
  const navigate = useAliasNavigate();
  const { pathname } = useAliasResolvedPath();

  return useCallback(
    (path, { isLeftNav = false, isPassSearchHash = true, ...options } = {}) => {
      if (isLeftNav) {
        return undefined;
      }

      const { firstMatch } = routerHelpers.getRouteConfigByPath({ pathName: path });

      if (firstMatch) {
        const dynamicBaseName = routerHelpers.dynamicBaseName({ pathName: pathname });
        const updatedPath = `${dynamicBaseName}/${firstMatch.productPath}`;

        return navigate((isPassSearchHash && `${updatedPath}${search}${hash}`) || updatedPath, options);
      }

      return navigate((isPassSearchHash && `${path}${search}${hash}`) || path, options);
    },
    [hash, navigate, pathname, search]
  );
};

/**
 * Search parameter, return
 *
 * @param {object} options
 * @param {Function} options.useLocation
 * @param {Function} options.useSearchParams
 * @returns {Array}
 */
const useSearchParams = ({
  useSearchParams: useAliasSearchParams = useRRDSearchParams,
  useLocation: useAliasLocation = useLocation
} = {}) => {
  const { search } = useAliasLocation();
  const [, setAliasSearchParams] = useAliasSearchParams();

  /**
   * Alias returned React Router Dom useSearchParams hook to something expected.
   * Defaults to merging search objects instead of overwriting them.
   *
   * @param {object} updatedQuery
   * @param {object} options
   * @param {boolean} options.isMerged Merge search with existing search, or don't
   * @param {string|*} options.currentSearch search returned from useLocation
   */
  const setSearchParams = (updatedQuery, { isMerged = true, currentSearch = search } = {}) => {
    let updatedSearch = {};

    if (isMerged) {
      Object.assign(updatedSearch, routerHelpers.parseSearchParams(currentSearch), updatedQuery);
    } else {
      updatedSearch = updatedQuery;
    }

    setAliasSearchParams(updatedSearch);
  };

  return [routerHelpers.parseSearchParams(search), setSearchParams];
};

const context = {
  useLocation,
  useNavigate,
  useParams,
  useRedirect,
  useRouteDetail,
  useSearchParams
};

export {
  context as default,
  context,
  useLocation,
  useNavigate,
  useParams,
  useRedirect,
  useRouteDetail,
  useSearchParams
};
