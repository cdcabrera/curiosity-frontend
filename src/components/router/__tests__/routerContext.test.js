// import { productConfig } from '../../../config';
import { context, useLocation, useNavigate, useRouteDetail } from '../routerContext';

describe('RouterContext', () => {
  it('should return specific properties', () => {
    expect(context).toMatchSnapshot('specific properties');
  });

  it('should apply a hook for return routeDetail', () => {
    const mockUseRouterContext = () => ({
      routeDetail: {
        lorem: 'ipsum'
      },
      dolor: 'sit'
    });

    const { result } = shallowHook(() => useRouteDetail({ useRouterContext: mockUseRouterContext }));
    expect(result).toMatchSnapshot('route details');
  });

  it('should apply a hook for useLocation', async () => {
    const mockLocation = {
      search: '?lorem=ipsum'
    };

    const { result: mockUseLocation } = await mountHook(() => useLocation({ useLocation: () => mockLocation }));
    expect(mockUseLocation).toMatchSnapshot('location');
  });

  it('should apply a hook for useNavigate', () => {
    const mockLocation = {
      search: '?lorem=ipsum'
    };
    const mockNavigation = jest.fn;
    const updatedCalls = [];
    // const mockNavFunc = jest.fn().mockImplementation(() => mockNavigation);
    const { result: mockNavigationSet } = shallowHook(() =>
      useNavigate({
        useLocation: () => mockLocation,
        useNavigate: () => value => updatedCalls.push(value) // jest.fn().mockImplementation(() => mockNavigation)
        // useNavigate: () => mockNavFunc
      })
    );

    mockNavigationSet('/dolor/sit');
    mockNavigationSet('rhel');
    mockNavigationSet('insights');

    // console.log('>>>>>> result', mockNavigationSet, typeof mockNavigationSet);
    console.log('>>>>>>>> result', mockNavigation);
    // /{
    //       rhods: mockNavigationSet('/dolor/sit'),
    //       rhel: [mockNavigationSet('rhel'), mockNavigationSet('insights')]
    //     }

    expect(updatedCalls).toMatchSnapshot('navigation push');
  });
});
