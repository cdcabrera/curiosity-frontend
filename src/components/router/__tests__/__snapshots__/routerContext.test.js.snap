// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouterContext should apply a hook for useLocation: location 1`] = `
{
  "_id": "c-679636502",
  "lorem": "ipsum",
  "updateLocation": [Function],
}
`;

exports[`RouterContext should apply a hook for useNavigate: navigation dispatch 1`] = `
[
  [
    {
      "config": "rhel",
      "type": "SET_PRODUCT",
    },
  ],
  [
    {
      "config": "rhel",
      "type": "SET_PRODUCT",
    },
  ],
  [
    {
      "config": "rhel",
      "type": "SET_PRODUCT",
    },
  ],
]
`;

exports[`RouterContext should apply a hook for useNavigate: navigation push 1`] = `
[
  [
    {},
    "",
    "/subscriptions/rhel?lorem=ipsum",
    undefined,
  ],
  [
    {},
    "",
    "/subscriptions/rhel?lorem=ipsum",
    undefined,
  ],
  [
    {},
    "",
    "/subscriptions/rhel?lorem=ipsum",
    undefined,
  ],
]
`;

exports[`RouterContext should apply a hook for useRouteDetail: document title 1`] = `
[
  [
    "t(curiosity-view.title, {"appName":"Subscriptions","context":"rhel"}) - Subscriptions",
    true,
  ],
]
`;

exports[`RouterContext should apply a hook for useRouteDetail: route details, closest 1`] = `
{
  "detailProps": [
    "_passed",
    "allConfigs",
    "availableVariants",
    "firstMatch",
    "errorRoute",
    "isClosest",
    "productGroup",
    "productConfig",
    "productPath",
    "productVariant",
    "disableIsClosest",
  ],
  "firstMatch": [
    "aliases",
    "productGroup",
    "productId",
    "productLabel",
    "productPath",
    "productDisplay",
    "viewId",
    "productVariants",
    "query",
    "graphTallyQuery",
    "inventoryGuestsQuery",
    "inventoryHostsQuery",
    "inventorySubscriptionsQuery",
    "initialGraphFilters",
    "initialGraphSettings",
    "initialGuestsFilters",
    "initialInventoryFilters",
    "initialInventorySettings",
    "initialSubscriptionsInventoryFilters",
    "initialToolbarFilters",
  ],
  "isClosest": true,
  "productGroup": "rhel",
  "productPath": "l",
}
`;

exports[`RouterContext should apply a hook for useRouteDetail: route details, match 1`] = `
{
  "detailProps": [
    "_passed",
    "allConfigs",
    "availableVariants",
    "firstMatch",
    "errorRoute",
    "isClosest",
    "productGroup",
    "productConfig",
    "productPath",
    "productVariant",
    "disableIsClosest",
  ],
  "firstMatch": [
    "aliases",
    "productGroup",
    "productId",
    "productLabel",
    "productPath",
    "productDisplay",
    "viewId",
    "productVariants",
    "query",
    "graphTallyQuery",
    "inventoryGuestsQuery",
    "inventoryHostsQuery",
    "inventorySubscriptionsQuery",
    "initialGraphFilters",
    "initialGraphSettings",
    "initialGuestsFilters",
    "initialInventoryFilters",
    "initialInventorySettings",
    "initialSubscriptionsInventoryFilters",
    "initialToolbarFilters",
  ],
  "isClosest": false,
  "productGroup": "rhel",
  "productPath": "rhel",
}
`;

exports[`RouterContext should apply a hook for useSearchParams: initial params 1`] = `
{
  "lorem": "ipsum",
}
`;

exports[`RouterContext should apply a hook for useSetRouteDetail: dispatch route config path 1`] = `
[
  [
    {
      "config": "dolor-sit",
      "type": "SET_PRODUCT",
    },
  ],
]
`;

exports[`RouterContext should return specific properties: specific properties 1`] = `
{
  "useLocation": [Function],
  "useNavigate": [Function],
  "useRouteDetail": [Function],
  "useSearchParams": [Function],
  "useSetRouteDetail": [Function],
}
`;
